From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Manish Bhardwaj <mbhardwaj@nvidia.com>
Date: Fri, 12 Sep 2025 08:23:40 +0000
Subject: [PATCH 105/243] NVIDIA: SAUCE: tegra_hv: doxygen comment fixes in hv
 drivers

- add doxygen comment fixes in hv driver
- remove below APIs since there is not client using this API.
  - tegra_hv_ivc_set_loopback
  - tegra_hv_ivc_write_poke
  - tegra_hv_ivc_read_peek

Jira ESLC-8346
Bug 5433066

Change-Id: Ida8a2eece1c8a471d2aa4f5c7294d5245f9fa810
Signed-off-by: Manish Bhardwaj <mbhardwaj@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3451303
Reviewed-by: svc-mobile-coverity <svc-mobile-coverity@nvidia.com>
Reviewed-by: svcacv <svcacv@nvidia.com>
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
Reviewed-by: Sandeep Trasi <strasi@nvidia.com>
Reviewed-by: Suresh Venkatachalam <skathirampat@nvidia.com>
(cherry picked from commit c9314b533e3c6729e3add69faaf3a781d10c1c7e)

diff --git a/drivers/virt/tegra/tegra_hv.c b/drivers/virt/tegra/tegra_hv.c
index fbd5c1cb075f..371f4ea4c58b 100644
--- a/drivers/virt/tegra/tegra_hv.c
+++ b/drivers/virt/tegra/tegra_hv.c
@@ -939,17 +939,6 @@ int tegra_hv_ivc_read_user(struct tegra_hv_ivc_cookie *ivck, void __user *buf, i
 }
 EXPORT_SYMBOL(tegra_hv_ivc_read_user);
 
-#ifndef CONFIG_KERNEL_BUILD_WITH_PROD_DEFCONFIG
-int tegra_hv_ivc_read_peek(struct tegra_hv_ivc_cookie *ivck, void *buf,
-			   int off, int count)
-{
-	struct hv_ivc *ivc = cookie_to_ivc_dev(ivck);
-
-	return tegra_ivc_read_peek(&ivc->ivc, NULL, buf, off, count);
-}
-EXPORT_SYMBOL(tegra_hv_ivc_read_peek);
-#endif
-
 int tegra_hv_ivc_can_read(struct tegra_hv_ivc_cookie *ivck)
 {
 	struct hv_ivc *ivc = cookie_to_ivc_dev(ivck);
diff --git a/include/soc/tegra/virt/hv-ivc.h b/include/soc/tegra/virt/hv-ivc.h
index 35fdb9876e1f..fa61b6223449 100644
--- a/include/soc/tegra/virt/hv-ivc.h
+++ b/include/soc/tegra/virt/hv-ivc.h
@@ -9,301 +9,686 @@
 #include <linux/of.h>
 #include <linux/version.h>
 
+/**
+ * @defgroup hypervisor_ivc_framework Hypervisor IVC Framework
+ * @{
+ */
+
+/** @brief structure representing ivc queue cookie in hypervisor driver */
 struct tegra_hv_ivc_cookie {
-	/* some fields that might be useful */
+	/** @brief irq linked to the ivc queue */
 	int irq;
+	/** @brief vmid of the peer */
 	int peer_vmid;
+	/** @brief number of frames in an ivc queue */
 	int nframes;
+	/** @brief ivc frame size*/
 	int frame_size;
-	uint32_t *notify_va; /* address used to notify end-point */
+	/** @brief address used to notify end-point */
+	uint32_t *notify_va;
 };
 
+/** @brief represents the ivc queue operations */
 struct tegra_hv_ivc_ops {
-	/* called when data are received */
+	/** @brief called when data are received */
 	void (*rx_rdy)(struct tegra_hv_ivc_cookie *ivck);
-	/* called when space is available to write data */
+	/** @brief called when space is available to write data */
 	void (*tx_rdy)(struct tegra_hv_ivc_cookie *ivck);
 };
 
+/** @brief structure representing mempool cookie in hypervisor driver */
 struct tegra_hv_ivm_cookie {
+	/** @brief mempool base ipa address */
 	uint64_t ipa;
+	/** @brief mempool size */
 	uint64_t size;
+	/** @brief vmid of the peer */
 	unsigned int peer_vmid;
+	/** @brief reserved */
 	void *reserved;
 };
 
+/**
+ * @brief          checks whether platform supports virtualization or not
+ *
+ * @retval         true if platform supports virtualization else false.
+ *
+ * @pre            Tegra hypervisor driver should have been initialized.
+ *
+ * @post           Client can take runtime decision in driver if client
+ *                 driver support native and virtualized environment.
+ *
+ * @usage          Call this function to check if OS is running in native environment
+ *                 or in virtualized environment.
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
+ */
 bool is_tegra_hypervisor_mode(void);
 
 /**
- * tegra_hv_ivc_reserve - Reserve an IVC queue for use
- * @dn:		Device node pointer to the queue in the DT
- *		If NULL, then operate on first HV device
- * @queue_id	Id number of the queue to use.
- * @ops		Ops structure or NULL (deprecated)
- *
- * Reserves the queue for use
- *
- * Returns a pointer to the ivc_dev to use or an ERR_PTR.
- * Note that returning EPROBE_DEFER means that the ivc driver
- * hasn't loaded yet and you should try again later in the
- * boot sequence.
- *
- * Note that @ops must be NULL for channels that handle reset.
+ * @brief          Reserve an IVC queue for use
+ * @param[in]      dn Device node pointer to the queue in the DT
+ *		   If NULL, then operate on first HV device
+ * @param[in]      id Id number of the queue to use.
+ * @param[in]      ops Ops structure or NULL (deprecated)
+ *
+ * @retval         ptr ivc queue cookie pointer or else errors.
+ *
+ * @pre            Tegra hypervisor driver should have been initialized.
+ *                 This API should be invoked on virtual/hypervisor environment only.
+ *
+ * @post           After reserving IVC queue client can perform
+ *                 I/O operation on the IVC queue.
+ *
+ * @usage          reserve ivc queue before performing I/O operations.
+ *                 driver should have been probed successfully.
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
  */
 struct tegra_hv_ivc_cookie *tegra_hv_ivc_reserve(
 		struct device_node *dn, int id,
 		const struct tegra_hv_ivc_ops *ops);
 
 /**
- * tegra_hv_ivc_unreserve - Unreserve an IVC queue used
- * @ivck	IVC cookie
- *
- * Unreserves the IVC channel
- *
- * Returns 0 on success and an error code otherwise
+ * @brief          Unreserve an IVC queue used
+ * @param[in]      ivck IVC cookie of the queue
+ *
+ * @retval         0 on success and an error code otherwise
+ *
+ * @pre            Tegra hypervisor driver should have been initialized.
+ *                 This API should be invoked on virtual/hypervisor environment only.
+ *
+ * @post           After unreserving IVC queue client can not perform
+ *                 I/O operation on the IVC queue.
+ *
+ * @usage          unreserve ivc queue after performing I/O operations.
+ *                 driver should have been probed successfully.
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
  */
 int tegra_hv_ivc_unreserve(struct tegra_hv_ivc_cookie *ivck);
 
 /**
- * ivc_hv_ivc_write - Writes a frame to the IVC queue
- * @ivck	IVC cookie of the queue
- * @buf		Pointer to the data to write
- * @size	Size of the data to write
- *
- * Write a number of bytes (as a single frame) from the queue.
- *
- * Returns size on success and an error code otherwise
+ * @brief          Write a number of bytes (as a single frame) from the queue.
+ * @param[in]      ivck IVC cookie of the queue
+ * @param[in]      buf Pointer to the data to write
+ * @param[in]      size Size of the data to write
+ *
+ * @retval         size on success and an error code otherwise
+ *
+ * @pre            Tegra hypervisor driver should have been initialized.
+ *                 This API should be invoked on virtual/hypervisor environment only.
+ *                 IVC should have been reserved before performing any ivc operations.
+ *
+ * @post           data will be written on the ivc queue successfully.
+ *
+ * @usage          use this API to send data to peer end.
+ *                 ivc should have been reserved before using this API.
+ *                 never use this API after unreserving ivc queue.
+ *                 driver should have been probed successfully.
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
  */
 int tegra_hv_ivc_write(struct tegra_hv_ivc_cookie *ivck, const void *buf,
 		int size);
 
 /**
- * ivc_hv_ivc_write_user - Writes a frame to the IVC queue
- * @ivck	IVC cookie of the queue
- * @buf		Pointer to the userspace data to write
- * @size	Size of the data to write
- *
- * Write a number of bytes (as a single frame) from the queue.
- *
- * Returns size on success and an error code otherwise
+ * @brief          Write a number of bytes (as a single frame) from the queue.
+ * @param[in]      ivck IVC cookie of the queue
+ * @param[in]      buf Pointer to the userspace data to write
+ * @param[in]      size Size of the data to write
+ *
+ * @retval         size Returns size on success and an error code otherwise
+ *
+ *
+ * @pre            Tegra hypervisor driver should have been initialized.
+ *                 This API should be invoked on virtual/hypervisor environment only.
+ *                 IVC should have been reserved before performing any ivc operations.
+ *
+ * @post           data will be written on the ivc queue successfully from userspace buffer.
+ *
+ * @usage          use this API to send data to peer end.
+ *                 ivc should have been reserved before using this API.
+ *                 never use this API after unreserving ivc queue.
+ *                 driver should have been probed successfully.
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
  */
 int tegra_hv_ivc_write_user(struct tegra_hv_ivc_cookie *ivck, const void __user *buf,
 		int size);
 
 /**
- * ivc_hv_ivc_read - Reads a frame from the IVC queue
- * @ivck	IVC cookie of the queue
- * @buf		Pointer to the data to read
- * @size	max size of the data to read
- *
- * Reads a number of bytes (as a single frame) from the queue.
- *
- * Returns size on success and an error code otherwise
+ * @brief          Reads a number of bytes (as a single frame) from the queue.
+ * @param[in]      ivck IVC cookie of the queue
+ * @param[in,out]  buf Pointer to the data to read
+ * @param[in]      size max size of the data to read
+ *
+ * @retval         size Returns size on success and an error code otherwise
+ *
+ * @pre            Tegra hypervisor driver should have been initialized.
+ *                 This API should be invoked on virtual/hypervisor environment only.
+ *                 IVC should have been reserved before performing any ivc operations.
+ *
+ * @post           data will be read from the ivc queue successfully.
+ *
+ * @usage          use this API to receive data from peer end.
+ *                 ivc should have been reserved before using this API.
+ *                 never use this API after unreserving ivc queue.
+ *                 driver should have been probed successfully.
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
  */
 int tegra_hv_ivc_read(struct tegra_hv_ivc_cookie *ivck, void *buf, int size);
 
 /**
- * ivc_hv_ivc_read_user - Reads a frame from the IVC queue
- * @ivck	IVC cookie of the queue
- * @buf		Pointer to the userspace data to read
- * @size	max size of the data to read
- *
- * Reads a number of bytes (as a single frame) from the queue.
- *
- * Returns size on success and an error code otherwise
+ * @brief          Reads a number of bytes (as a single frame) from the queue.
+ * @param[in]      ivck IVC cookie of the queue
+ * @param[in,out]  buf Pointer to the userspace data to read
+ * @param[in]      size max size of the data to read
+ *
+ * @retval         size Returns size on success and an error code otherwise
+ *
+ * @pre            Tegra hypervisor driver should have been initialized.
+ *                 This API should be invoked on virtual/hypervisor environment only.
+ *                 IVC should have been reserved before performing any ivc operations.
+ *
+ * @post           data will be read from the ivc queue successfully into userspace buffer.
+ *
+ * @usage          use this API to receive data from peer end.
+ *                 ivc should have been reserved before using this API.
+ *                 never use this API after unreserving ivc queue.
+ *                 driver should have been probed successfully.
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
  */
 int tegra_hv_ivc_read_user(struct tegra_hv_ivc_cookie *ivck, void __user *buf, int size);
 
 /**
- * ivc_hv_ivc_can_read - Test whether data are available
- * @ivck	IVC cookie of the queue
- *
- * Test whether data to read are available
- *
- * Returns 1 if data are available in the rx queue, 0 if not
+ * @brief          Test whether data are available
+ * @param[in]      ivck IVC cookie of the queue
+ *
+ * @retval         1 if data are available in the rx queue, 0 if not
+ *
+ * @pre            Tegra hypervisor driver should have been initialized.
+ *                 This API should be invoked on virtual/hypervisor environment only.
+ *                 IVC should have been reserved before performing any ivc operations.
+ *
+ * @post           client take decision to read data if data is present
+ *
+ * @usage          use this API if client want to check if response has been received peer end.
+ *                 ivc should have been reserved before using this API.
+ *                 never use this API after unreserving ivc queue.
+ *                 driver should have been probed successfully.
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
  */
 int tegra_hv_ivc_can_read(struct tegra_hv_ivc_cookie *ivck);
 
 /**
- * ivc_hv_ivc_can_write - Test whether data can be written
- * @ivck	IVC cookie of the queue
- *
- * Test whether data can be written
- *
- * Returns 1 if data are can be written to the tx queue, 0 if not
+ * @brief          Test whether data can be written
+ * @param[in]      ivck IVC cookie of the queue
+ *
+ * @retval         1 if data are can be written to the tx queue, 0 if not
+ *
+ * @pre            Tegra hypervisor driver should have been initialized.
+ *                 This API should be invoked on virtual/hypervisor environment only.
+ *                 IVC should have been reserved before performing any ivc operations.
+ *
+ * @post           client take decision to write data if write slot is free.
+ *
+ * @usage          use this API if client want to check if write slot is present or not.
+ *                 ivc should have been reserved before using this API.
+ *                 never use this API after unreserving ivc queue.
+ *                 driver should have been probed successfully.
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
  */
 int tegra_hv_ivc_can_write(struct tegra_hv_ivc_cookie *ivck);
 
 /**
- * ivc_hv_ivc_tx_empty - Test whether the tx queue is empty
- * @ivck	IVC cookie of the queue
- *
- * Test whether the tx queue is completely empty
- *
- * Returns 1 if the queue is empty, zero otherwise
+ * @brief          Test whether the tx queue is empty
+ * @param[in]      ivck IVC cookie of the queue
+ *
+ * @retval         1 if the queue is empty, zero otherwise
+ *
+ * @pre            Tegra hypervisor driver should have been initialized.
+ *                 This API should be invoked on virtual/hypervisor environment only.
+ *                 IVC should have been reserved before performing any ivc operations.
+ *
+ * @post           client will know whether tx queue is empty or not.
+ *
+ * @usage          use this API if tx queue is empty or not.
+ *                 ivc should have been reserved before using this API.
+ *                 never use this API after unreserving ivc queue.
+ *                 driver should have been probed successfully.
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
  */
 int tegra_hv_ivc_tx_empty(struct tegra_hv_ivc_cookie *ivck);
 
 #ifndef CONFIG_KERNEL_BUILD_WITH_PROD_DEFCONFIG
-/**
- * tegra_ivc_tx_frames_available - gets number of free entries in tx queue
- * @ivc/@ivck	IVC channel or cookie
- *
- * Returns the number of unused entries in the tx queue. Assuming the caller
- * does not write any additional frames, this number may increase from the
- * value returned as the receiver consumes frames.
- *
+/*
+ * @brief          gets number of free entries in tx queue
+ *                 Returns the number of unused entries in the tx queue. Assuming the caller
+ *                 does not write any additional frames, this number may increase from the
+ *                 value returned as the receiver consumes frames.
+ * @param[in]      ivck IVC cookie of the queue
+ *
+ * @retval         num Number ot frames available.
+ *
+ * @pre            Tegra hypervisor driver should have been initialized.
+ *                 This API should be invoked on virtual/hypervisor environment only.
+ *                 IVC should have been reserved before performing any ivc operations.
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
  */
 uint32_t tegra_hv_ivc_tx_frames_available(struct tegra_hv_ivc_cookie *ivck);
 
-/**
- * ivc_hv_ivc_loopback - Sets (or clears) loopback mode
- * @ivck	IVC cookie of the queue
- * @mode	Set loopback on/off (1 = on, 0 = off)
- *
- * Sets or clears loopback mode accordingly.
- *
- * When loopback is active any writes are ignored, while
- * reads do not return data.
- * Incoming data are copied immediately to the tx queue.
- *
- * Returns 0 on success, a negative error code otherwise
+/*
+ * @brief          Dump ivc info in dmesg logs
+ * @param[in]      ivck IVC cookie of the queue
+ *
+ * @retval         0 on success, a negative error code otherwise
+ *
+ * @pre            Tegra hypervisor driver should have been initialized.
+ *                 This API should be invoked on virtual/hypervisor environment only.
+ *                 IVC should have been reserved before performing any ivc operations.
+ * @usage
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
  */
-int tegra_hv_ivc_set_loopback(struct tegra_hv_ivc_cookie *ivck, int mode);
-
-/* debugging aid */
 int tegra_hv_ivc_dump(struct tegra_hv_ivc_cookie *ivck);
-
-/**
- * ivc_hv_ivc_read_peek - Peek (copying) data from a received frame
- * @ivck	IVC cookie of the queue
- * @buf		Buffer to receive the data
- * @off		Offset in the frame
- * @count	Count of bytes to copy
- *
- * Peek data from a received frame, copying to buf, without removing
- * the frame from the queue.
- *
- * Returns 0 on success, a negative error code otherwise
- */
-int tegra_hv_ivc_read_peek(struct tegra_hv_ivc_cookie *ivck,
-		void *buf, int off, int count);
 #endif
 
 /**
- * ivc_hv_ivc_read_get_next_frame - Peek at the next frame to receive
- * @ivck	IVC cookie of the queue
- *
- * Peek at the next frame to be received, without removing it from
- * the queue.
- *
- * Returns a pointer to the frame, or an error encoded pointer.
+ * @brief          Peek at the next frame to receive
+ *                 Peek at the next frame to be received, without removing it from
+ *                 the queue.
+ *
+ * @param[in]      ivck IVC cookie of the queue
+ *
+ * @retval         ptr Returns a pointer to the frame, or an error encoded pointer.
+ *
+ * @pre            Tegra hypervisor driver should have been initialized.
+ *                 This API should be invoked on virtual/hypervisor environment only.
+ *                 IVC should have been reserved before performing any ivc operations.
+ *
+ * @post           after peeking next frame we can check the content of frames.
+ *
+ * @usage          if client want to check next frame without removing it from queue.
+ *                 ivc should have been reserved before using this API.
+ *                 never use this API after unreserving ivc queue.
+ *                 driver should have been probed successfully.
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
  */
 void *tegra_hv_ivc_read_get_next_frame(struct tegra_hv_ivc_cookie *ivck);
 
 /**
- * ivc_hv_ivc_read_advance - Advance the read queue
- * @ivck	IVC cookie of the queue
- *
- * Advance the read queue
- *
- * Returns 0, or a negative error value if failed.
+ * @brief          Advance the read queue
+ * @param[in]      ivck IVC cookie of the queue
+ *
+ * @retval         0, or a negative error value if failed.
+ *
+ * @pre            Tegra hypervisor driver should have been initialized.
+ *                 This API should be invoked on virtual/hypervisor environment only.
+ *                 IVC should have been reserved before performing any ivc operations.
+ *
+ * @post           internal ivc counters point to next read frame.
+ *
+ * @usage          use this API if client wants to advance to next read frame
+ *                 ivc should have been reserved before using this API.
+ *                 never use this API after unreserving ivc queue.
+ *                 driver should have been probed successfully.
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
  */
 int tegra_hv_ivc_read_advance(struct tegra_hv_ivc_cookie *ivck);
 
 /**
- * ivc_hv_ivc_write_poke - Poke data to a frame to be transmitted
- * @ivck	IVC cookie of the queue
- * @buf		Buffer to the data
- * @off		Offset in the frame
- * @count	Count of bytes to copy
- *
- * Copy data to a transmit frame, copying from buf, without advancing
- * the transmit queue.
- *
- * Returns 0 on success, a negative error code otherwise
- */
-int tegra_hv_ivc_write_poke(struct tegra_hv_ivc_cookie *ivck,
-		const void *buf, int off, int count);
-
-/**
- * ivc_hv_ivc_write_get_next_frame - Poke at the next frame to transmit
- * @ivck	IVC cookie of the queue
- *
- * Get access to the next frame.
- *
- * Returns a pointer to the frame, or an error encoded pointer.
+ * @brief          Poke at the next frame to transmit
+ *                 Get access to the next frame.
+ * @param[in]      ivck IVC cookie of the queue
+ *
+ * @retval         ptr Returns a pointer to the frame, or an error encoded pointer.
+ *
+ * @pre            Tegra hypervisor driver should have been initialized.
+ *                 This API should be invoked on virtual/hypervisor environment only.
+ *                 IVC should have been reserved before performing any ivc operations.
+ *
+ * @post           client can write data to frame after getting it without advacing tx frame counter.
+ *
+ * @usage          if client want to send data to peer end without removing it from queue.
+ *                 ivc should have been reserved before using this API.
+ *                 never use this API after unreserving ivc queue.
+ *                 driver should have been probed successfully.
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
  */
 void *tegra_hv_ivc_write_get_next_frame(struct tegra_hv_ivc_cookie *ivck);
 
 /**
- * ivc_hv_ivc_write_advance - Advance the write queue
- * @ivck	IVC cookie of the queue
- *
- * Advance the write queue
- *
- * Returns 0, or a negative error value if failed.
+ * @brief          Advance the write queue
+ * @param[in]      ivck IVC cookie of the queue
+ *
+ * @retval         0 on success or a negative error value if failed.
+ *
+ * @pre            Tegra hypervisor driver should have been initialized.
+ *                 This API should be invoked on virtual/hypervisor environment only.
+ *                 IVC should have been reserved before performing any ivc operations.
+ *
+ * @post           internal ivc counters point to next write frame.
+ *
+ * @usage          use this API if client wants to advance to next write frame
+ *                 ivc should have been reserved before using this API.
+ *                 never use this API after unreserving ivc queue.
+ *                 driver should have been probed successfully.
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
  */
 int tegra_hv_ivc_write_advance(struct tegra_hv_ivc_cookie *ivck);
 
 /**
- * tegra_hv_mempool_reserve - reserve a mempool for use
- * @id		Id of the requested mempool.
- *
- * Returns a cookie representing the mempool on success, otherwise an ERR_PTR.
+ * @brief          reserve a mempool for use
+ * @param[in]      id Id of the requested mempool.
+ *
+ * @retval         ivck Returns a cookie representing the mempool on success, otherwise an ERR_PTR.
+ *
+ * @pre            Tegra hypervisor driver should have been initialized.
+ *                 This API should be invoked on virtual/hypervisor environment only.
+ *
+ * @post           reserved mempool will be available for I/O operations.
+ *
+ * @usage          mempool will be available for data transfer with peer end.
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
  */
 struct tegra_hv_ivm_cookie *tegra_hv_mempool_reserve(unsigned int id);
 
 /**
- * tegra_hv_mempool_release - release a reserved mempool
- * @ck		Cookie returned by tegra_hv_mempool_reserve().
- *
- * Returns 0 on success or a negative error code otherwise.
+ * @brief          release a reserved mempool
+ * @param[in]      ivck IVC cookie of the queue
+ *
+ * @retval         ret 0 on success or a negative error code otherwise.
+ *
+ * @pre            Tegra hypervisor driver should have been initialized.
+ *                 This API should be invoked on virtual/hypervisor environment only.
+ *
+ * @post           unreserved mempool will not be available for I/O operations.
+ *
+ * @usage          mempool will not be available for data transfer available for data transfer.
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
  */
-int tegra_hv_mempool_unreserve(struct tegra_hv_ivm_cookie *ck);
+int tegra_hv_mempool_unreserve(struct tegra_hv_ivm_cookie *ivck);
 
 /**
- * ivc_channel_notified - handle internal messages
- * @ivck	IVC cookie of the queue
- *
- * This function must be called following every notification (interrupt or
- * callback invocation) for the tegra_hv_- version).
- *
- * Returns 0 if the channel is ready for communication, or -EAGAIN if a channel
- * reset is in progress.
+ * @brief          handle internal messages
+ *                 This function must be called following every notification (interrupt or
+ *                 callback invocation) for the tegra_hv_- version).
+ * @param[in]      ivck IVC cookie of the queue
+ *
+ * @retval         0 if the channel is ready for communication, or -EAGAIN if a channel
+ *                 reset is in progress.
+ *
+ * @pre            Tegra hypervisor driver should have been initialized.
+ *                 This API should be invoked on virtual/hypervisor environment only.
+ *                 IVC should have been reserved before performing any ivc operations.
+ *
+ * @post           IVC channel will be establised with peer end and will be ready for communication
+ *
+ * @usage          use this APi to establish connection with peer end for communication
+ *                 ivc should have been reserved before using this API.
+ *                 never use this API after unreserving ivc queue.
+ *                 driver should have been probed successfully.
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
  */
 int tegra_hv_ivc_channel_notified(struct tegra_hv_ivc_cookie *ivck);
 
 /**
- * ivc_channel_reset - initiates a reset of the shared memory state
- * @ivck	IVC cookie of the queue
- *
- * This function must be called after a channel is reserved before it is used
- * for communication. The channel will be ready for use when a subsequent call
- * to ivc_channel_notified() returns 0.
+ * @brief          initiates a reset of the shared memory state
+ *                 This function must be called after a channel is reserved before it is used
+ *                 for communication. The channel will be ready for use when a subsequent call
+ *                 to ivc_channel_notified() returns 0.
+ * @param[in]      ivck IVC cookie of the queue
+ *
+ * @retval         None
+ *
+ * @pre            Tegra hypervisor driver should have been initialized.
+ *                 This API should be invoked on virtual/hypervisor environment only.
+ *                 IVC should have been reserved before performing any ivc operations.
+ *
+ * @post           channel will be in reseted state.
+ *
+ * @usage          before establishing connection reset ivc channel first after reserving ivc queue.
+ *                 ivc should have been reserved before using this API.
+ *                 never use this API after unreserving ivc queue.
+ *                 driver should have been probed successfully.
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
  */
 void tegra_hv_ivc_channel_reset(struct tegra_hv_ivc_cookie *ivck);
 
 /**
- * tegra_hv_ivc_get_info - Get info of Guest shared area
- * @ivck	IVC cookie of the queue
- * @pa		IPA of shared area
- * @size	Size of the shared area
- *
- * Get info (IPA and size) of Guest shared area
- *
- * Returns size on success and an error code otherwise
+ * @brief          Get info (IPA and size) of Guest shared area
+ * @param[in]      ivck IVC cookie of the queue
+ * @param[out]     pa IPA of shared area
+ * @param[out]     size Size of the shared area
+ *
+ * @retval         0 on success & -EINVAL on failure.
+ *
+ * @pre            Tegra hypervisor driver should have been initialized.
+ *                 This API should be invoked on virtual/hypervisor environment only.
+ *                 IVC should have been reserved before performing any ivc operations.
+ *
+ * @post           IPA address and size of ivc queue will be available to client.
+ *
+ * @usage          when client want to mmap ivc memory to userspace via mmap API.
+ *                 driver should have been probed successfully.
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
  */
 int tegra_hv_ivc_get_info(struct tegra_hv_ivc_cookie *ivck, uint64_t *pa,
 			  uint64_t *size);
 
 /**
- * tegra_hv_ivc_notify - Notify remote guest
- * @ivck	IVC cookie of the queue
- *
- * Notify remote guest
- *
+ * @brief          Notify remote guest
+ * @param[in]      ivck IVC cookie of the queue
+ *
+ * @retval         None
+ *
+ * @pre            Tegra hypervisor driver should have been initialized.
+ *                 This API should be invoked on virtual/hypervisor environment only.
+ *                 IVC should have been reserved before performing any ivc operations.
+ *
+ * @post           remote guest will be notified
+ *
+ * @usage          use when want to notify the peer end after writing or reading data to ivc queue
+ *                 ivc should have been reserved before using this API.
+ *                 never use this API after unreserving ivc queue.
+ *                 driver should have been probed successfully.
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
  */
 void tegra_hv_ivc_notify(struct tegra_hv_ivc_cookie *ivck);
 
+/** @} */
+
 #endif /* __TEGRA_HV_IVC_H */
diff --git a/include/soc/tegra/virt/syscalls.h b/include/soc/tegra/virt/syscalls.h
index 03f97c46976a..0497f749d4a9 100644
--- a/include/soc/tegra/virt/syscalls.h
+++ b/include/soc/tegra/virt/syscalls.h
@@ -8,34 +8,84 @@
 
 #include <soc/tegra/virt/tegra_hv_sysmgr.h>
 
+/**
+ * @defgroup hypervisor_calls Hypervisor Calls
+ * @{
+ */
+
+/* @brief `always_inline` to ensure it is always inlined
+ *          for performance-critical operations.
+ */
+#define ALWAYS_INLINE			__attribute__((always_inline)) inline
+
+/*
+ * @brief `no_sanitize_address` attribute to avoid
+ * address sanitizer instrumentation, necessary for low-level code.
+ */
+#define NO_SANITIZE_ADDRESS		__attribute__((no_sanitize_address))
+
+/* @brief Immediate hypercall value used to get read stat */
 #define HVC_NR_READ_STAT		1
-#define HVC_NR_READ_IVC			2
-#define HVC_NR_READ_GID			3
-#define HVC_NR_RAISE_IRQ		4
-#define HVC_NR_READ_NGUESTS		5
-#define HVC_NR_READ_IPA_PA		6
-#define HVC_NR_READ_GUEST_STATE		7
-#define HVC_NR_READ_HYP_INFO		9
-#define HVC_NR_GUEST_RESET		10
-#define HVC_NR_SYSINFO_IPA		13
-#define HVC_NR_LCPU0_MPIDR		15
-#define HVC_NR_READ_VM_INFO		16
-#define HVC_NR_TRACE_GET_EVENT_DATA	0x8003U
-#define HVC_NR_TRACE_SET_EVENT_DATA	0x8004U
+/* @brief Immediate hypercall value used in the hyp_read_ivc_info() wrapper */
+#define HVC_NR_READ_IVC			2u
+/** @brief Immediate hypercall value used in the hyp_read_gid() wrapper */
+#define HVC_NR_READ_GID			3u
+/* @brief Immediate hypercall value used in the hyp_raise_irq() wrapper */
+#define HVC_NR_RAISE_IRQ		4u
+/* @brief Immediate hypercall value used in the hyp_read_nguests() wrapper */
+#define HVC_NR_READ_NGUESTS		5u
+/* @brief Immediate hypercall value used in the hyp_read_ipa_pa_info() wrapper */
+#define HVC_NR_READ_IPA_PA		6u
+/* @brief Immediate hypercall value used in the hyp_read_guest_state() wrapper */
+#define HVC_NR_READ_GUEST_STATE		7u
+/* @brief Immediate hypercall value used in the hyp_read_hyp_info() wrapper */
+#define HVC_NR_READ_HYP_INFO		9u
+/* @brief Immediate hypercall value used in the hyp_guest_reset() wrapper */
+#define HVC_NR_GUEST_RESET		10u
+/* @brief Immediate hypercall value used in the hyp_sysinfo_ipa() wrapper */
+#define HVC_NR_SYSINFO_IPA		13u
+/* @brief Immediate hypercall value used in the hyp_lcpu0_mpidr() wrapper */
+#define HVC_NR_LCPU0_MPIDR		15u
+/* @brief Immediate hypercall value used in the hyp_read_vm_info() wrapper */
+#define HVC_NR_READ_VM_INFO		16u
+/* @brief Immediate hypercall value used in the hyp_trace_get_mask() wrapper */
+#define HVC_NR_TRACE_GET_EVENT_DATA		0x8003U
+/* @brief Immediate hypercall value used in the hyp_trace_set_mask() wrapper */
+#define HVC_NR_TRACE_SET_EVENT_DATA		0x8004U
+/* @brief Immediate hypercall value used in the hyp_smmu_diagnostic() wrapper */
 #define HVC_NR_SMMU_DIAG		0x8009U
-#define HVC_NR_CBB_DIAG			0x8133U
-#define HVC_NR_MC_DIAG			0x8134U
+/* @brief Immediate hypercall value used in the hyp_cbb_err_diagnostic() wrapper */
+#define HVC_NR_CBB_DIAG		0x8133U
+/* @brief Immediate hypercall value used in the hyp_mc_err_diagnostic() wrapper */
+#define HVC_NR_MC_DIAG		0x8134U
 
+/* @brief FIXME Not being Used */
 #define GUEST_PRIMARY		0
+/* @brief FIXME Not being Used */
 #define GUEST_IVC_SERVER	0
+/* @brief Max number of nvlog producers */
 #define MAX_NVLOG_PRODUCERS	32U
+/* @brief Immediate hypercall value used in
+ * - hyp_read_freq_feedback() wrapper
+ * - hyp_read_freq_request() wrapper
+ * - hyp_write_freq_request() wrapper
+ * - hyp_pct_cpu_id_read_freq_feedback() wrapper
+ * - hyp_pct_cpu_id_read_freq_request() wrapper
+ * - hyp_pct_cpu_id_write_freq_request() wrapper
+ * - hyp_get_cpu_count() wrapper
+ */
 #define HVC_NR_CPU_FREQ		0xC6000022
 
+/* @brief Set masks to dump evenlib events */
 #define TRACE_SET_EVENT_MASK		0x0U
+/* @brief Get masks to dump evenlib events */
 #define TRACE_GET_EVENT_MASK		0x1U
+/* @brief Set the hypervisor profiler frequency. */
 #define TRACE_SET_PROFILER_FREQ	 	0x2U
+/* @brief Get the hypervisor profiler frequency. */
 #define TRACE_GET_PROFILER_FREQ	 	0x3U
 
+/* @brief Max number of VM supported by Hypervisor */
 #define NGUESTS_MAX 16
 
 #ifndef __ASSEMBLY__
@@ -44,36 +94,36 @@
 #include <linux/types.h>
 #endif
 
-/*
- * @brief Structure describing a single NvLog producer. A producer can have
+/* @brief
+ * Structure describing a single NvLog producer. A producer can have
  * multiple buffers (if it has multiple threads) and the buffers are stored
  * contiguously in memory as an array. This structure provides the base
  * address, stride, and length of the array so that the consumer can locate all
  * of the buffers belonging to the producer.
  */
 struct nvlog_producer {
-	/* Base IPA of an array of NvLog buffers belonging to the producer. */
+	/* @brief Base IPA of an array of NvLog buffers belonging to the producer. */
 	uint64_t ipa;
-	/* Size of the IPA region containing the NvLog buffer array. */
+	/* @brief Size of the IPA region containing the NvLog buffer array. */
 	uint64_t region_size;
-	/* Size of a single NvLog buffer. This is the stride of the array of
+	/* @brief Size of a single NvLog buffer. This is the stride of the array of
 	 * NvLog buffers belonging to the producer.
 	 */
 	uint64_t buf_size;
-	/* Number of NvLog buffers belonging to the producer. */
+	/* @brief Number of NvLog buffers belonging to the producer. */
 	uint64_t buf_count;
-	/* Name of the NvLog producer. */
+	/* @brief Name of the NvLog producer. */
 	char name[32];
 };
 
-/*
+/* @brief
  * Data structure for the VM Information Region.
  *
- * The Intermediate Physical Address (IPA) of this structure is returned by the
+ * The Intermediate Physical Address (IPA) of this structure is @returned by the
  * hyp_read_vm_info Hypercall.
  */
 struct vm_info_region {
-	/*
+	/* @brief
 	 * Table of NvLog producers.
 	 *
 	 * This table is only populated for VMs that have the 'log_access' PCT flag
@@ -83,66 +133,151 @@ struct vm_info_region {
 	struct nvlog_producer nvlog_producers[MAX_NVLOG_PRODUCERS];
 };
 
+/**
+ * @brief Data structure for an IVC queue.
+ * Although an IVC queue points to the total shared memory that is required for
+ * communication between RX and TX, the size field here only indicates the size
+ * of RX FIFO or the TX FIFO which is half the size of the total IVC Queue
+ * memory. This is to support backward compatibility with the Linux
+ * drivers as well as the Hypervisor that maps this shared memory for each end
+ * to use.
+ *
+ * @note Structure should be 8 bytes aligned, specified by compiler directives
+ * for performance and atomicity guarantees.
+ */
 struct tegra_hv_queue_data {
-	uint32_t	id;	/* IVC id */
+	/* @brief IVC queue id */
+	uint32_t	id;
+	/* @brief IDs of the two users of the queue */
 	uint32_t	peers[2];
+	/* @brief size of SIVC FIFO memory (data plus header) */
 	uint32_t	size;
+	/* @brief number of data frames in each direction */
 	uint32_t	nframes;
+	/* @brief size of each frame in bytes */
 	uint32_t	frame_size;
+	/* @brief offset of this structure from the start of the IVC area */
 	uint32_t	offset;
-	uint16_t	irq, raise_irq;
-	uint64_t	trap_ipa; /** @brief IO address used to notify peer endpoint */
-	uint64_t	msi_ipa; /** @brief MSI address used to notify peer endpoint */
+	/* @brief virtual IRQ ID associated to this queue */
+	uint16_t	irq;
+	/* @brief reserved for internal use */
+	uint16_t	raise_irq;
+	/* @brief IO address used to notify peer endpoint */
+	uint64_t	trap_ipa;
+	/* @brief MSI address used to notify peer endpoint */
+	uint64_t	msi_ipa;
 };
 
+/**
+ * @brief Data structure for an IVC mempool.
+ *
+ * @note Structure should be 8 bytes aligned, specified by compiler directives
+ * for performance and atomicity guarantees.
+ */
 struct ivc_mempool {
+	/* @brief start address of the mempool shared memory */
 	uint64_t pa;
+	/* @brief size of the mempool shared memory in bytes */
 	uint64_t size;
+	/* @brief ID of one of the mempool's clients */
 	uint32_t id;
+	/* @brief ID of the other mempool client */
 	uint32_t peer_vmid;
 };
 
+/**
+ * @brief Data structure for an IVC shared area.
+ *
+ * @note Structure should be 8 bytes aligned, specified by compiler directives
+ * for performance and atomicity guarantees.
+ */
 struct ivc_shared_area {
+	/* @brief start address of the shared area */
 	uint64_t pa;
+	/* @brief size of the area in bytes */
 	uint64_t size;
+	/* @brief ID of the VM to which this area belongs */
 	uint32_t guest;
+	/* @brief starting index of free irqs */
 	uint16_t free_irq_start;
+	/* @brief total count of free irqs */
 	uint16_t free_irq_count;
 };
 
+/**
+ * @brief Data structure that contains information about all IVC objects (queues
+ * and mempools) used by the current VM.
+ *
+ * @note Structure should be 8 bytes aligned, specified by compiler directives
+ * for performance and atomicity guarantees.
+ */
 struct ivc_info_page {
+	/* @brief Total number of queues */
 	uint32_t nr_queues;
+	/* @brief Total number of areas */
 	uint32_t nr_areas;
+	/* @brief Total number of mempools */
 	uint32_t nr_mempools;
-	uint32_t padding; /**< @brief reserved for internal use */
-			// IMPORTANT: Padding is needed to align
-			// sizeof(struct ivc_info_page ) to 64 bits
-	uint64_t trap_region_base_ipa; /**< @brief MMIO trap region start address */
-	uint64_t trap_region_size; /**< @brief MMIO trap region size */
-	uint64_t trap_ipa_stride; /**< @brief MMIO trap IPA stride size */
-	uint64_t msi_region_base_ipa; /**< @brief MMIO msi region start address */
-	uint64_t msi_region_size; /**< @brief MMIO msi region size */
-	uint64_t msi_ipa_stride; /**< @brief MMIO msi IPA stride size */
-
-	/* The actual length of this array is nr_areas. */
+	/* @brief reserved for internal use
+	 *
+	 * @note IMPORTANT: Padding is needed to align
+	 * sizeof(struct ivc_info_page ) to 64 bits
+	 */
+	uint32_t padding;
+	/* @brief MMIO trap region start address */
+	uint64_t trap_region_base_ipa;
+	/* @brief MMIO trap region size */
+	uint64_t trap_region_size;
+	/* @brief MMIO trap IPA stride size */
+	uint64_t trap_ipa_stride;
+	/* @brief MMIO msi region start address */
+	uint64_t msi_region_base_ipa;
+	/* @brief MMIO msi region size */
+	uint64_t msi_region_size;
+	/* @brief MMIO msi IPA stride size */
+	uint64_t msi_ipa_stride;
+	/* @brief The actual length of this array is nr_areas. */
 	struct ivc_shared_area areas[];
 
-	/*
-	 * Following the shared array is an array of queue data structures with
-	 * an entry per queue that is assigned to the guest. This  array is
-	 * terminated by an entry with no frames.
-	 *
+	/**
+	 * The IVC shared area follows the above fields.
+	 * Accessing the IVC shared area must be done via the inline function
+	 * ivc_shared_area_addr().
+	 * The IVC shared area consists of an array of nr_areas ivc_shared areas
+	 * structures: struct ivc_shared_area areas[nr_areas];
+	 */
+
+	/**
+	 * Following the IVC shared area is the array of queue data structures
+	 * with one entry per each queue that is accessible in the current VM.
+	 * Accessing the queue data structures must be done via the inline
+	 * function ivc_info_queue_array().
+	 * The queue data area consists of an array of nr_queues
+	 * tegra_hv_queue_data structures:
 	 * struct tegra_hv_queue_data queue_data[nr_queues];
 	 */
 
-	/*
-	 * Following the queue data array is an array of mempool structures
-	 * with an entry per mempool assigned to the guest.
-	 *
-	 * struct ivc_mempool[nr_mempools];
+	/**
+	 * Following the queue data area is the array of mempool structures
+	 * with one entry per each mempool that is accessible in the current VM.
+	 * Accessing the mempool data structures must be done via the inline
+	 * function ivc_info_mempool_array().
+	 * The mempool data area consists of an array of nr_mempools
+	 * ivc_mempool structures:
+	 * struct ivc_mempool mempool_data[nr_mempools];
 	 */
 };
 
+/*
+ * @brief          Obtain pointer to the shared area represnted by area number
+ *
+ * @param[in]      info IVC info page where IVC meta data is maintained by Hypervisor
+ * @param[in]      area_num represents Area Number
+ *
+ * @return         @returns a pointer to the associated ivc_shared_area structure
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
 static inline struct ivc_shared_area *ivc_shared_area_addr(
 		const struct ivc_info_page *info, uint32_t area_num)
 {
@@ -150,12 +285,30 @@ static inline struct ivc_shared_area *ivc_shared_area_addr(
 		+ area_num;
 }
 
+/*
+ * @brief          Obtain pointer to the array of IVC Queue
+ *
+ * @param[in]      info IVC info page where IVC meta data is maintained by Hypervisor
+ *
+ * @return         pointer to the array of IVC Queue
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
 static inline const struct tegra_hv_queue_data *ivc_info_queue_array(
 		const struct ivc_info_page *info)
 {
 	return (struct tegra_hv_queue_data *)&info->areas[info->nr_areas];
 }
 
+/*
+ * @brief          Obtain pointer to the array of mempool
+ *
+ * @param[in]      info IVC info page where IVC meta data is maintained by Hypervisor
+ *
+ * @return         pointer to the array of mempool
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
 static inline const struct ivc_mempool *ivc_info_mempool_array(
 		const struct ivc_info_page *info)
 {
@@ -163,77 +316,129 @@ static inline const struct ivc_mempool *ivc_info_mempool_array(
 			&ivc_info_queue_array(info)[info->nr_queues];
 }
 
+/* @brief Data structure used in the hyp_read_ipa_pa_info interface */
 struct hyp_ipa_pa_info {
-	uint64_t base;       /* base of contiguous pa region */
-	uint64_t offset;     /* offset for requested ipa address */
-	uint64_t size;       /* size of pa region */
+	/* @brief base of contiguous pa region */
+	uint64_t base;
+	/* @brief offset for requested ipa address */
+	uint64_t offset;
+	/* @brief size of pa region */
+	uint64_t size;
 };
 
+/* @brief Max VCPUs supported by Hypervisor */
 #define HVC_MAX_VCPU 64
 
+/* @brief FIXME */
 struct trapped_access {
+	/* @brief FIXME */
 	uint64_t ipa;
+	/* @brief FIXME */
 	uint32_t size;
+	/* @brief FIXME */
 	int32_t write_not_read;
+	/* @brief FIXME */
 	uint64_t data;
+	/* @brief FIXME */
 	uint32_t guest_id;
 };
 
-/* Structure to store the IPA, Length and Name. */
+/* @brief @brief Structure to store the IPA, Length and Name. */
 struct trace_buf {
-/* @brief IPA of trace buffer region. */
+	/* @brief IPA of trace buffer region. */
 	uint64_t ipa;
-/* @brief Length of trace buffer region. */
+	/* @brief Length of trace buffer region. */
 	uint64_t size;
-/* @brief Name */
+	/* @brief Trace Buffer Name */
 	char name[32];
 };
 
+/*
+ * @brief Data structure for the server's information page.
+ *
+ * The Intermediate Physical Address (IPA) of this structure is
+ * @returned by the hyp_read_hyp_info Hypercall.
+ */
 struct hyp_server_page {
-	/* guest reset protocol */
+	/* @brief Virtual IRQ index used in the guest reset protocol */
 	uint32_t guest_reset_virq;
 
-	/* boot delay offsets per VM needed by monitor partition */
+	/* @brief boot delay offsets per VM needed by monitor partition */
 	uint32_t boot_delay[NGUESTS_MAX];
 
-	/* hypervisor trace log */
+	/* @brief The IPA of the Hypervisor trace log buffer. */
 	uint64_t log_ipa;
+	/* @brief The size of the Hypervisor trace log buffer. */
 	uint32_t log_size;
 
-	/* secure-hypervisor trace log */
+	/* @brief IPA of Hypervisor secure trace log buffer. */
 	uint64_t secure_log_ipa;
+	/* @brief Size of Hypervisor secure trace log buffer. */
 	uint32_t secure_log_size;
 
-	/* PCT data */
+	/* @brief The IPA of a the PCT blob, 0 if the VM is not a server. */
 	uint64_t pct_ipa;
+	/* @brief The size of the PCT blob. */
 	uint64_t pct_size;
 
-	/* check if the VM is a server or a guest */
+	/* @brief Indicates whether the VM is a server.
+	 *
+	 * It is non-zero if the VM is a server, zero otherwise.
+	 */
 	uint32_t is_server_vm;
 
-	/* golden register data */
+	/* @brief The IPA of the Golden Register data area */
 	uint64_t gr_ipa;
+	/* @brief The size of the Golden Register data area */
 	uint32_t gr_size;
 
-	/* all vm mappings ipa */
+	/* @brief all vm mappings ipa */
 	uint64_t mappings_ipa;
 
-	/* IPA, Length and Name for trace buffer region. */
+	/* @brief IPA, Length and Name trace buffer region. */
 	struct trace_buf trace_buffs[2*NGUESTS_MAX];
 
 };
 
-/* For backwards compatibility, alias the old name for hyp_server_name. */
+/* @brief For backwards compatibility, alias the old name for hyp_server_name. */
 #define hyp_info_page hyp_server_page
 
+/* @brief CPU core registers from x3 to x17 */
 #define _X3_X17 "x3", "x4", "x5", "x6", "x7", "x8", "x9", "x10", "x11", "x12", \
 "x13", "x14", "x15", "x16", "x17"
 
+/* @brief CPU core registers from x4 to x17 */
 #define _X4_X17 "x4", "x5", "x6", "x7", "x8", "x9", "x10", "x11", "x12", \
 "x13", "x14", "x15", "x16", "x17"
 
-__attribute__((no_sanitize_address))
-static inline int hyp_read_gid(unsigned int *gid)
+/**
+ * @brief          Obtain the ID of the current VM.
+ *
+ * @param[in,out]  gid pointer to an int location.
+ *                 Index of current VM in struct guest_conf [] arrary (a.k.a PCT VM Index)
+ *                 is updated in the provided location on successful @return from this API.
+ * @retval         0 if successful.
+ * @retval         -1 if the input pointer (gid) points to NULL.
+ *
+ * @invariant      Pointer gid shall point to a vaild location.
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ *
+ * @post           Guest VM ID will be returned.
+ *
+ * @usage
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: Yes
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
+ */
+NO_SANITIZE_ADDRESS static inline int hyp_read_gid(unsigned int *gid)
 {
 	register uint64_t r0 asm("x0");
 	register uint64_t r1 asm("x1");
@@ -247,8 +452,20 @@ static inline int hyp_read_gid(unsigned int *gid)
 	return (int)r0;
 }
 
-__attribute__((no_sanitize_address))
-static inline int hyp_read_nguests(unsigned int *nguests)
+/*
+ * @brief          Obtain the number of Virtual Machines in the current configuration.
+ *
+ * @param[in,out]  nguests pointer to an int location
+ *
+ * @return         0 if successful: *nguests will contain the number of VMs;
+ *                 non-zero otherwise
+ *
+ * @note           Only server VMs are allowed to obtain this information; for
+ *                 non-server VMs this function @returns an error value.
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
+NO_SANITIZE_ADDRESS static inline int hyp_read_nguests(unsigned int *nguests)
 {
 	register uint64_t r0 asm("x0");
 	register uint64_t r1 asm("x1");
@@ -277,8 +494,23 @@ static inline int hyp_read_vm_info(uint64_t *vm_info_region_pa)
 	return (int)r0;
 }
 
-__attribute__((no_sanitize_address))
-static inline int hyp_read_ivc_info(uint64_t *ivc_info_page_pa)
+/*
+ * @brief          Obtain Intermediate Physical Address of the location containing
+ *	           the VMs IVC & Mempool information (struct ivc_info_page).
+ *
+ * @param[in,out]  ivc_info_page_pa pointer to the location where the Intermediate
+ *	           Physical Address of the Page (64K in size) containing VMs IVC and Mempool
+ *	           information (refer struct ivc_info_page) is updated on successful @return.
+ *
+ * @return         Function success or error code.
+ * @retval         0 if successful
+ * @retval         -IVC_EINVAL if ivc_info_page_pa value is NULL.
+ *
+ * @invarinat      ivc_info_page_pa pointer shall point to a vaild location.
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
+NO_SANITIZE_ADDRESS static inline int hyp_read_ivc_info(uint64_t *ivc_info_page_pa)
 {
 	register uint64_t r0 asm("x0");
 	register uint64_t r1 asm("x1");
@@ -292,8 +524,28 @@ static inline int hyp_read_ivc_info(uint64_t *ivc_info_page_pa)
 	return (int)r0;
 }
 
-__attribute__((no_sanitize_address))
-static inline int hyp_read_ipa_pa_info(struct hyp_ipa_pa_info *info, unsigned int guestid, uint64_t ipa)
+/*
+ * @brief          Obtain contiguous physical block (start & size) that corresponds
+ *                 to the given Guest Intermediate Physical Address.
+ *
+ * @param[in,out]  info struct hyp_ipa_pa_info* const
+ *	           [in] pointer to struct hyp_ipa_pa_info where the physical block information
+ *	           is @returned. -IVC_EINVAL is @returned in case this pointer is NULL
+ *	           [out] struct hyp_ipa_pa_info* const info In case of successful @return value for this interface,
+ *	           struct hyp_ipa_pa_info pointer location is updated as below:
+ *	           struct hyp_ipa_pa_info.base contains start of contiguous physical block in which input IPA lies
+ *	           struct hyp_ipa_pa_info.offset contains the offset with-in the contiguous physical block of the input IPA
+ *	           struct hyp_ipa_pa_info.size contains the size of the contiguous physical block in which input IPA lies.
+ * @param[in]      guestid PCT index of the guest for which the physical block information is to be obtained.
+ * @param[in]      ipa IPA for which the physical contiguous block information is to be obtained.
+ *
+ * @return         Whether requested information was successfully obtained.
+ * @retval         r0 value will be @return by function.
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
+NO_SANITIZE_ADDRESS static inline int hyp_read_ipa_pa_info(struct hyp_ipa_pa_info *info,
+		unsigned int guestid, uint64_t ipa)
 {
 	register uint64_t r0 asm("x0") = guestid;
 	register uint64_t r1 asm("x1") = ipa;
@@ -313,8 +565,18 @@ static inline int hyp_read_ipa_pa_info(struct hyp_ipa_pa_info *info, unsigned in
 	return (int)r0;
 }
 
-__attribute__((no_sanitize_address))
-static inline int hyp_raise_irq(unsigned int irq, unsigned int vmid)
+/*
+ * @brief          Obtain the state of a given VM.
+ *
+ * @param[in]      irq to be raised
+ * @param[in]      vmid the ID of the VM
+ *
+ * @return         0 if successful; -IVC_EINVAL otherwise.
+ * @return         Value @return by Hypervisor in r0 register
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
+NO_SANITIZE_ADDRESS static inline int hyp_raise_irq(unsigned int irq, unsigned int vmid)
 {
 	register uint64_t r0 asm("x0") = irq;
 	register uint64_t r1 asm("x1") = vmid;
@@ -327,8 +589,19 @@ static inline int hyp_raise_irq(unsigned int irq, unsigned int vmid)
 	return (int)r0;
 }
 
-__attribute__((no_sanitize_address))
-static inline int hyp_read_guest_state(unsigned int vmid, unsigned int *state)
+/*
+ * @brief          Obtain the state of a given VM.
+ *
+ * @param[in]      vmid the ID of the VM
+ * @param[in,out]  state pointer to an int location
+ *
+ * @return         0 if successful; -IVC_EINVAL otherwise.
+ *                 On successful completion, *state will be filled with the
+ *                 state value as defined by the enumeration vm_state.
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
+NO_SANITIZE_ADDRESS static inline int hyp_read_guest_state(unsigned int vmid, unsigned int *state)
 {
 	register uint64_t r0 asm("x0") = vmid;
 	register uint64_t r1 asm("x1");
@@ -342,8 +615,26 @@ static inline int hyp_read_guest_state(unsigned int vmid, unsigned int *state)
 	return (int)r0;
 }
 
-__attribute__((no_sanitize_address))
-static inline int hyp_read_hyp_info(uint64_t *hyp_info_page_pa)
+/*
+ * @brief          Obtain the Intermediate Physical Address of the Server Page for the
+ *                 current server VM.
+ *
+ * @param[in,out]  hyp_info_page_pa pointer to a uint64_t location. On successful
+ *                 completion, *hyp_info_page_pa will contain the IPA of the
+ *                 current VM's Server Page.
+ *
+ * @return         Whether the IPA was obtained successfully.
+ * @retval         0 for success.
+ * @retval         -IVC_EINVAL if hyp_info_page_pa is NULL.
+ * @retval         -EINVAL if function is called by non-server VM (identified by
+ *                 struct guest_conf->is_server_guest == 0).
+ *
+ * @note           Only server VMs have a Server Page; for non-server VMs this
+ *                 function @returns an error value -EINVAL.
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
+NO_SANITIZE_ADDRESS static inline int hyp_read_hyp_info(uint64_t *hyp_info_page_pa)
 {
 	register uint64_t r0 asm("x0");
 	register uint64_t r1 asm("x1");
@@ -357,8 +648,26 @@ static inline int hyp_read_hyp_info(uint64_t *hyp_info_page_pa)
 	return (int)r0;
 }
 
-__attribute__((no_sanitize_address))
-static inline int hyp_guest_reset(unsigned int id, struct hyp_sys_state_info *out)
+/*
+ * @brief          Initiate a System state transition from a privileged VM.
+ *
+ * @param[in]      id The system state transition Command ID which can be one of following:
+ *                 - SYS_SHUTDOWN_INIT_CMD : Indicates System Shutdown. Refer to @p hyp_sys_shutdown for ICD.
+ *                 - SYS_REBOOT_INIT_CMD   : Indicates System Reboot. Refer to @p hyp_sys_reboot for ICD.
+ *                 - SYS_SUSPEND_INIT_CMD  : Indicates SC7. Refer to @p hyp_sys_suspend for ICD.
+ *                 - GUEST_PAUSE_CMD       : Indicates Guest Suspend. Refer to @p hyp_guest_pause for ICD.
+ * @param[out]     out is not used.
+ *
+ * @return         int         The function @returns an integer value.
+ * @retval         0           System State Transition was successfully requested.
+ * @retval         -EINVAL     if input command id is not supported.
+ * @retval         -EACCESS    if guest VM issuing the call is not privileged. A privileged VM is a VM that
+ *                 has field "is_privileged" set to 1U in guestCfg structure from PCT.
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
+NO_SANITIZE_ADDRESS static inline int hyp_guest_reset(unsigned int id,
+				  struct hyp_sys_state_info *out)
 {
 	register uint64_t r0 asm("x0") = id;
 	register uint64_t r1 asm("x1");
@@ -394,8 +703,14 @@ static inline int hyp_guest_enter_vm_op(unsigned int const id, unsigned int cons
 	return (int)r0;
 }
 
-__attribute__((no_sanitize_address))
-static inline uint64_t hyp_sysinfo_ipa(void)
+/*
+ * @brief          Return the IPA of the SysInfo area for the current VM.
+ *
+ * @return         non-zero IPA address
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
+NO_SANITIZE_ADDRESS static inline uint64_t hyp_sysinfo_ipa(void)
 {
 	register uint64_t r0 asm("x0");
 
@@ -407,8 +722,17 @@ static inline uint64_t hyp_sysinfo_ipa(void)
 	return r0;
 }
 
-__attribute__((no_sanitize_address))
-static inline int hyp_read_freq_feedback(uint64_t *value)
+/*
+ * @brief          API to read frequency feedback
+ *
+ * @param[in,out]  value pointer to location as input where result will be updated.
+ *                 Return value from r1 register will be store in value pointer
+ *
+ * @return         Value @return by Hypervisor in r0 register
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
+NO_SANITIZE_ADDRESS static inline int hyp_read_freq_feedback(uint64_t *value)
 {
 	register uint64_t r0 asm("x0") = HVC_NR_CPU_FREQ;
 	register uint64_t r1 asm("x1") = 1U;
@@ -424,8 +748,17 @@ static inline int hyp_read_freq_feedback(uint64_t *value)
 	return (int16_t)r0;
 }
 
-__attribute__((no_sanitize_address))
-static inline int hyp_read_freq_request(uint64_t *value)
+/*
+ * @brief          API to read frequency request
+ *
+ * @param[in,out]  value pointer to location as input where result will be updated.
+ *                 Return value from r1 register will be store in value pointer
+ *
+ * @return         Value @return by Hypervisor in r0 register
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
+NO_SANITIZE_ADDRESS static inline int hyp_read_freq_request(uint64_t *value)
 {
 	register uint64_t r0 asm("x0") = HVC_NR_CPU_FREQ;
 	register uint64_t r1 asm("x1") = 0U;
@@ -441,8 +774,16 @@ static inline int hyp_read_freq_request(uint64_t *value)
 	return (int16_t)r0;
 }
 
-__attribute__((no_sanitize_address))
-static inline int hyp_write_freq_request(uint64_t value)
+/*
+ * @brief          API to write frequency request
+ *
+ * @param[in]      value frequency value to be set.
+ *
+ * @return         Value @return by Hypervisor in r0 register
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
+NO_SANITIZE_ADDRESS static inline int hyp_write_freq_request(uint64_t value)
 {
 	register uint64_t r0 asm("x0") = HVC_NR_CPU_FREQ;
 	register uint64_t r1 asm("x1") = 2U;
@@ -456,8 +797,19 @@ static inline int hyp_write_freq_request(uint64_t value)
 	return (int16_t)r0;
 }
 
-__attribute__((no_sanitize_address))
-static inline int hyp_pct_cpu_id_read_freq_feedback(uint8_t cpu_id, uint64_t *value)
+/*
+ * @brief          API to read frequency feedback for dedicated CPU
+ *
+ * @param[in]      cpu_id ID to the dedicated CPU for which freq to be read.
+ * @param[in,out]  value pointer to location as input where result will be updated.
+ *                 Return value from r1 register will be store in value pointer
+ *
+ * @return         Value @return by Hypervisor in r0 register
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
+NO_SANITIZE_ADDRESS static inline int hyp_pct_cpu_id_read_freq_feedback(uint8_t cpu_id,
+							uint64_t *value)
 {
 	register uint64_t r0 asm("x0") = HVC_NR_CPU_FREQ;
 	register uint64_t r1 asm("x1") = 4U;
@@ -475,8 +827,19 @@ static inline int hyp_pct_cpu_id_read_freq_feedback(uint8_t cpu_id, uint64_t *va
 
 }
 
-__attribute__((no_sanitize_address))
-static inline int hyp_pct_cpu_id_read_freq_request(uint8_t cpu_id, uint64_t *value)
+/*
+ * @brief          API to read frequency request for dedicated CPU
+ *
+ * @param[in]      cpu_id ID to the dedicated CPU for which freq to be read.
+ * @param[in,out]  value pointer to location as input where result will be updated.
+ *                 Return value from r1 register will be store in value pointer
+ *
+ * @return         Value @return by Hypervisor in r0 register
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
+NO_SANITIZE_ADDRESS static inline int hyp_pct_cpu_id_read_freq_request(uint8_t cpu_id,
+							uint64_t *value)
 {
 	register uint64_t r0 asm("x0") = HVC_NR_CPU_FREQ;
 	register uint64_t r1 asm("x1") = 3U;
@@ -493,8 +856,18 @@ static inline int hyp_pct_cpu_id_read_freq_request(uint8_t cpu_id, uint64_t *val
 	return (int16_t)r0;
 }
 
-__attribute__((no_sanitize_address))
-static inline int hyp_pct_cpu_id_write_freq_request(uint8_t cpu_id, uint64_t value)
+/*
+ * @brief          API to write frequency request for dedicated CPU
+ *
+ * @param[in]      cpu_id CPU id for which freq is written
+ * @param[in]      value frequency value to be written
+ *
+ * @return         Value @return by Hypervisor in r0 register
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
+NO_SANITIZE_ADDRESS static inline int hyp_pct_cpu_id_write_freq_request(uint8_t cpu_id,
+							uint64_t value)
 {
 	register uint64_t r0 asm("x0") = HVC_NR_CPU_FREQ;
 	register uint64_t r1 asm("x1") = 5U;
@@ -509,8 +882,14 @@ static inline int hyp_pct_cpu_id_write_freq_request(uint8_t cpu_id, uint64_t val
 	return (int16_t)r0;
 }
 
-__attribute__((no_sanitize_address))
-static inline uint8_t hyp_get_cpu_count(void)
+/*
+ * @brief          API to get CPU count
+ *
+ * @return         Value @return by Hypervisor in r1 register
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
+NO_SANITIZE_ADDRESS static inline uint8_t hyp_get_cpu_count(void)
 {
 	register uint64_t r0 asm("x0") = HVC_NR_CPU_FREQ;
 	register uint64_t r1 asm("x1") = 6U;
@@ -526,8 +905,19 @@ static inline uint8_t hyp_get_cpu_count(void)
 	return 0;
 }
 
-__attribute__((no_sanitize_address))
-static __attribute__((always_inline)) inline void hyp_call44(uint16_t id, uint64_t args[4])
+/*
+ * @brief          API to trigger hypervisor call
+ *
+ * @param[in,out]  id Hypervisor Call id for which HVC call is being invoked
+ * @param[in]      args pointer to the array of arguments of size 4.
+ *                 args[4] gets updated with stored value of x0 to x3 registers
+ *
+ * @return         Value @return by Hypervisor in r0 register
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
+NO_SANITIZE_ADDRESS static ALWAYS_INLINE inline void hyp_call44(uint16_t id,
+			uint64_t *args)
 {
 		register uint64_t x0 asm("x0") = args[0];
 		register uint64_t x1 asm("x1") = args[1];
@@ -545,8 +935,17 @@ static __attribute__((always_inline)) inline void hyp_call44(uint16_t id, uint64
 		args[3] = x3;
 }
 
-__attribute__((no_sanitize_address))
-static inline int hyp_trace_get_mask(uint64_t *value)
+/*
+ * @brief          Get mask set for hypervisor trace buffers
+ *
+ * @param[in,out]  value pointer will be passed as input.
+ *                 Return value from r1 register will be store in value pointer
+ *
+ * @return         Value @return by Hypervisor in r0 register
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
+NO_SANITIZE_ADDRESS static inline int hyp_trace_get_mask(uint64_t *value)
 {
 	uint64_t args[4] = { TRACE_GET_EVENT_MASK, 0U, 0U, 0U };
 	hyp_call44(HVC_NR_TRACE_GET_EVENT_DATA, args);
@@ -556,16 +955,36 @@ static inline int hyp_trace_get_mask(uint64_t *value)
 	return (int) args[0];
 }
 
-__attribute__((no_sanitize_address))
-static inline int hyp_trace_set_mask(uint64_t mask)
+/*
+ * @brief          Set mask for hypervisor trace buffers
+ *
+ * @param[in]      mask variable will be passed as input.
+ *
+ * @return         Value @return by Hypervisor in r0 register
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
+NO_SANITIZE_ADDRESS static inline int hyp_trace_set_mask(uint64_t mask)
 {
 	uint64_t args[4] = { TRACE_SET_EVENT_MASK, mask, 0U, 0U};
 	hyp_call44(HVC_NR_TRACE_SET_EVENT_DATA, args);
 	return (int) args[0];
 }
 
-__attribute__((no_sanitize_address))
-static inline int hyp_trace_get_profiler_freq(uint64_t *freq)
+/*
+ * @brief          Get the current hypervisor profiler frequency.
+ *
+ * @param[out]     freq Pointer to a uint64_t location where the current frequency
+ *                 will be stored on success.
+ *
+ * @return         Function success or error code.
+ * @retval         0 if successful.
+ * @retval         -IVC_EINVAL if the input pointer (freq) is NULL.
+ * @retval         Other non-zero value indicates a hypervisor error (e.g., permission denied).
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
+NO_SANITIZE_ADDRESS static inline int hyp_trace_get_profiler_freq(uint64_t *freq)
 {
 	uint64_t args[4] = { TRACE_GET_PROFILER_FREQ, 0U, 0U, 0U };
 	hyp_call44(HVC_NR_TRACE_GET_EVENT_DATA, args);
@@ -575,16 +994,39 @@ static inline int hyp_trace_get_profiler_freq(uint64_t *freq)
 	return (int) args[0];
 }
 
-__attribute__((no_sanitize_address))
-static inline int hyp_trace_set_profiler_freq(uint64_t freq)
+/*
+ * @brief          Set the hypervisor profiler frequency.
+ *
+ * @param[in]      freq The frequency value to set.
+ *
+ * @return         Function success or error code.
+ * @retval         0 if successful.
+ * @retval         Other non-zero value indicates a hypervisor error (e.g., permission denied).
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
+NO_SANITIZE_ADDRESS static inline int hyp_trace_set_profiler_freq(uint64_t freq)
 {
 	uint64_t args[4] = { TRACE_SET_PROFILER_FREQ, freq, 0U, 0U};
 	hyp_call44(HVC_NR_TRACE_SET_EVENT_DATA, args);
 	return (int) args[0];
 }
 
-__attribute__((no_sanitize_address))
-static inline bool hyp_smmu_diagnostic(void)
+/*
+ * @brief          Perform SMMU diagnostics.
+ *
+ * @details        This interface detects and reports SMMU errors that has happened
+ *                 prior to calling this interface.
+ *                 For all the SMMU instances this interface checks for any global fault, and context fault.
+ *                 If there is/are any then those are reported along with the fault details read from fault
+ *                 syndrome registers, such as the type of faults, fault address, Stream Id etc.
+ *
+ * @return         bool type that is based on the value held in X0 after HVC call is complete.
+ *                 It @returns true if X0 holds 0x1UL, false in case X0 holds 0x0UL.
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
+NO_SANITIZE_ADDRESS static inline bool hyp_smmu_diagnostic(void)
 {
 	uint64_t args[4] = { 0U, 0U, 0U, 0U };
 	hyp_call44(HVC_NR_SMMU_DIAG, args);
@@ -592,8 +1034,18 @@ static inline bool hyp_smmu_diagnostic(void)
 	return (args[0] == 0U);
 }
 
-__attribute__((no_sanitize_address))
-static inline bool hyp_cbb_err_diagnostic(void)
+/*
+ * @brief          Perform CBB diagnostics.
+ *
+ * @details        This interface detects and report CBB errors that has happened
+ *                 prior to calling this interface.
+ *
+ * @return         bool type that is based on the value held in X0 after HVC call is complete.
+ *                 It @returns true if X0 holds 0x1UL, false in case X0 holds 0x0UL.
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
+NO_SANITIZE_ADDRESS static inline bool hyp_cbb_err_diagnostic(void)
 {
 	uint64_t args[4] = { 0U, 0U, 0U, 0U };
 	hyp_call44(HVC_NR_CBB_DIAG, args);
@@ -601,8 +1053,18 @@ static inline bool hyp_cbb_err_diagnostic(void)
 	return (args[0] == 0U);
 }
 
-__attribute__((no_sanitize_address))
-static inline bool hyp_mc_err_diagnostic(void)
+/*
+ * @brief          Perform MC diagnostics.
+ *
+ * @details        This interface detects and report MC errors that has happened
+ *                 prior to calling this interface.
+ *
+ * @return         bool type that is based on the value held in X0 after HVC call is complete.
+ *                 It @returns true if X0 holds 0x1UL, false in case X0 holds 0x0UL.
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
+NO_SANITIZE_ADDRESS static inline bool hyp_mc_err_diagnostic(void)
 {
 	uint64_t args[4] = { 0U, 0U, 0U, 0U };
 	hyp_call44(HVC_NR_MC_DIAG, args);
@@ -610,8 +1072,18 @@ static inline bool hyp_mc_err_diagnostic(void)
 	return (args[0] == 0U);
 }
 
-__attribute__((no_sanitize_address))
-static inline long int hyp_lcpu0_mpidr(void)
+/*
+ * @brief          Return the MPIDR value for LCPU0
+ *
+ * @details        This API @returns the MPIDR_EL1 value of physical (host) CPU0.  The @return value is
+ *                 constant across multiple calls while the system remains powered on.
+ *
+ * @return         mpidr value for LCPU0
+ * @note           The caller of this API is not required to have a VCPU that runs on LCPU0.
+ *
+ * @pre            This API should be invoked on virtual/hypervisor environment only.
+ */
+NO_SANITIZE_ADDRESS static inline long int hyp_lcpu0_mpidr(void)
 {
 	uint64_t args[4] = { 0U, 0U, 0U, 0U };
 	hyp_call44(HVC_NR_LCPU0_MPIDR, args);
@@ -624,4 +1096,6 @@ static inline long int hyp_lcpu0_mpidr(void)
 
 #endif /* !__ASSEMBLY__ */
 
+/** @} */
+
 #endif /* __TEGRA_SYSCALLS_H__ */
diff --git a/include/soc/tegra/virt/tegra_hv.h b/include/soc/tegra/virt/tegra_hv.h
index 00c4fcb0e5db..b23dfae84a63 100644
--- a/include/soc/tegra/virt/tegra_hv.h
+++ b/include/soc/tegra/virt/tegra_hv.h
@@ -6,10 +6,17 @@
 #ifndef __TEGRA_HV_H__
 #define __TEGRA_HV_H__
 
+/**
+ * @defgroup hypervisor_ivc_framework Hypervisor IVC Framework
+ * @{
+ */
+
 #include <soc/tegra/virt/syscalls.h>
 
+/** @brief Supports trap notification to peer end via MSI irqs */
 #define SUPPORTS_TRAP_MSI_NOTIFICATION
 
+/** @brief Page size of IVC meta data queried from hypervisor */
 #define IVC_INFO_PAGE_SIZE 65536
 
 /** @brief Maximum Guest VM count */
@@ -19,7 +26,55 @@
 /** @brief The maximum number of mempools supported by the PCT. */
 #define PCT_MAX_NUM_MEMPOOLS	120U
 
+/**
+ * @brief          Query IVC meta data from hypervisor
+ *
+ * @retval         ptr to ivc_info_page structure having ivc meta data info
+ *                 shared by hypervisor
+ *
+ * @pre            Tegra hypervisor driver should have been initialized.
+ *                 This API should be invoked on virtual/hypervisor environment only.
+ *
+ * @post           IVC meta data will be retuned to client
+ *
+ * @usage
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: Yes
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
+ */
 const struct ivc_info_page *tegra_hv_get_ivc_info(void);
+
+/**
+ * @brief          Query vmid of the GOS VM from hypervisor
+ *
+ * @retval         vmid on success else -ve value
+ *
+ * @pre            Tegra hypervisor driver should have been initialized.
+ *                 This API should be invoked on virtual/hypervisor environment only.
+ *
+ * @post           Guest VM ID will be returned.
+ *
+ * @usage
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: Yes
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
+ */
 int tegra_hv_get_vmid(void);
 
+/** @} */
+
 #endif /* __TEGRA_HV_H__ */
diff --git a/include/soc/tegra/virt/tegra_hv_sysmgr.h b/include/soc/tegra/virt/tegra_hv_sysmgr.h
index d8ac0cc25552..8d15bc3e2e91 100644
--- a/include/soc/tegra/virt/tegra_hv_sysmgr.h
+++ b/include/soc/tegra/virt/tegra_hv_sysmgr.h
@@ -8,129 +8,207 @@
 
 #include <linux/types.h>
 
+/**
+ * @defgroup tegra_hv_sysmgr Hypervisor System Manager
+ * @{
+ */
+
+/** @brief Max size of IVC Message supported by IVC Queue */
 #define SYSMGR_IVCMSG_SIZE_MAX 64
 
+/** @brief Represents System Manager Message Types */
 enum hv_sysmgr_msg_type {
+	/** @brief System Manager Message representing Guest Event */
 	HV_SYSMGR_MSG_TYPE_GUEST_EVENT		= 1,
+	/** @brief System Manager Message representing VM PM Ctl Event */
 	HV_SYSMGR_MSG_TYPE_VM_PM_CTL_CMD	= 2,
+	/** @brief System Manager Message representing Invalid Event */
 	HV_SYSMGR_MSG_TYPE_INVALID
 };
 
+/** @brief Represents System Manager Request Command Ids */
 enum hv_sysmgr_cmd_id {
+	/** @brief System Manager Shutdown Command Id */
 	HV_SYSMGR_CMD_NORMAL_SHUTDOWN	= 0x0,
+	/** @brief System Manager Reboot Command Id */
 	HV_SYSMGR_CMD_NORMAL_REBOOT	= 0x1,
+	/** @brief System Manager Suspend Command Id */
 	HV_SYSMGR_CMD_NORMAL_SUSPEND	= 0x2,
+	/** @brief System Manager Resume Command Id */
 	HV_SYSMGR_CMD_NORMAL_RESUME	= 0x3,
+	/** @brief System Manager Invalid Command Id */
 	HV_SYSMGR_CMD_INVALID		= 0xFFFFFFFF,
 };
 
+/** @brief Represents System Manager Response Command Ids */
 enum hv_sysmgr_resp_id {
+	/** @brief System Manager Accepted Command Id */
 	HV_SYSMGR_RESP_ACCEPTED		= 0x0,
+	/** @brief System Manager Unknown Command Id */
 	HV_SYSMGR_RESP_UNKNOWN_COMMAND	= 0xF,
 };
 
-/* This struct comes as payload of hv_pm_ctl_message */
-struct hv_sysmgr_command {
+/**
+ * @brief This struct comes as payload of hv_pm_ctl_message.
+ * The __packed attribute ensures no padding is added.
+ */
+struct __packed hv_sysmgr_command {
+	/** @brief Command Id */
 	uint32_t cmd_id;
+	/** @brief Response Id */
 	uint32_t resp_id;
-} __packed;
+};
 
-struct hv_sysmgr_message {
-	/* msg class */
+/**
+ * @brief System Manager Message structure send or receive over IVC Queue.
+ * The __packed attribute ensures no padding is added.
+ */
+struct __packed hv_sysmgr_message {
+	/** @brief Message Type from hv_sysmgr_msg_type structure */
 	uint32_t msg_type;
-	/* id of open socket */
+	/** @brief Hypervisor sets to an opaque cookie */
 	uint32_t socket_id;
-	/* client data area. Payload */
+	/** @brief client data area. Payload of type hv_sysmgr_command structure */
 	uint8_t client_data[SYSMGR_IVCMSG_SIZE_MAX];
-} __packed;
+};
 
 
-/*
- * QUERY_SYSTEM_STATE COMMAND DATA LAYOUT
- */
+/** @brief QUERY_SYSTEM_STATE COMMAND DATA LAYOUT */
 struct hyp_sys_state_info {
-	/* Indicates System State Transition */
+	/** @brief Indicates System State Transition */
 	uint32_t sys_transition_mask;
 
-	/* Indicates which VM shutdown request is pending */
+	/** @brief Indicates which VM shutdown request is pending */
 	uint32_t vm_shutdown_mask;
 
-	/* Indicates which VM reboot request is pending */
+	/** @brief Indicates which VM reboot request is pending */
 	uint32_t vm_reboot_mask;
 
-	/* Indicates which VM suspend request is pending */
+	/** @brief Indicates which VM suspend request is pending */
 	uint32_t vm_suspend_phase_1_mask;
+	/** @brief Indicates which VM suspend request is pending */
 	uint32_t vm_suspend_phase_2_mask;
 
-	/* Indicates which VM resume request is pending */
+	/** @brief Indicates which VM resume request is pending */
 	uint32_t vm_resume_mask;
 };
 
-/*
- * Power management calls ID's used by SYSMGR to manage LOCAL/GLOBAL EVENTS
- */
+/** @brief Power management calls ID's used by SYSMGR to manage LOCAL/GLOBAL EVENTS */
 enum system_function_id {
+	/** @brief Represents Invalid Function */
 	INVALID_FUNC,
-	/*
+	/** @brief
 	 * This is used to get reboot/shutdown masks per VM from hypervisor.
 	 * Hypervisor updates state fields on a PSCI event from the VM.
 	 */
 	QUERY_SYSTEM_STATE,
+	/** @brief Represents Guest Shutdown Init State */
 	GUEST_SHUTDOWN_INIT,
+	/** @brief Represents Guest Shutdown Complete State */
 	GUEST_SHUTDOWN_COMPLETE,
+	/** @brief Represents Guest Reboot Init State */
 	GUEST_REBOOT_INIT,
+	/** @brief Represents Guest Reboot Continue State */
 	GUEST_REBOOT_CONTINUE,
+	/** @brief Represents Guest Reboot Complete State */
 	GUEST_REBOOT_COMPLETE,
+	/** @brief Represents System Shutdown Init State */
 	SYSTEM_SHUTDOWN_INIT,
+	/** @brief Represents System Shutdown Complete State */
 	SYSTEM_SHUTDOWN_COMPLETE,
+	/** @brief Represents System Reboot Init State */
 	SYSTEM_REBOOT_INIT,
+	/** @brief Represents System Reboot Complete State */
 	SYSTEM_REBOOT_COMPLETE,
+	/** @brief Represents Guest Suspend Request State */
 	GUEST_SUSPEND_REQ,
+	/** @brief Represents Guest Suspend Init State */
 	GUEST_SUSPEND_INIT,
+	/** @brief Represents Guest Suspend Complete State */
 	GUEST_SUSPEND_COMPLETE,
+	/** @brief Represents Guest Resume Init State */
 	GUEST_RESUME_INIT,
+	/** @brief Represents Guest Resume Complete State */
 	GUEST_RESUME_COMPLETE,
+	/** @brief Represents Guest Pause State */
 	GUEST_PAUSE,
+	/** @brief Represents System Suspend Init State */
 	SYSTEM_SUSPEND_INIT,
+	/** @brief Represents System Suspend Complete State */
 	SYSTEM_SUSPEND_COMPLETE,
 	GUEST_ENTER_VM_OP,
+	/** @brief Represents Max Possible Function Id */
 	MAX_FUNC_ID,
 };
 
+/** @brief Represents VM state and send to System Manager to maintian VM state */
 typedef enum  {
+	/** @brief Represents VM state as Booted */
 	VM_STATE_BOOT,
+	/** @brief Represents VM state as Halted */
 	VM_STATE_HALT,
+	/** @brief Represents VM state as Unhalted */
 	VM_STATE_UNHALT,
+	/** @brief Represents VM state as Reboted */
 	VM_STATE_REBOOT,
+	/** @brief Represents VM state as Shutdown */
 	VM_STATE_SHUTDOWN,
+	/** @brief Represents VM state as Suspend */
 	VM_STATE_SUSPEND,
+	/** @brief Represents VM state as Resume */
 	VM_STATE_RESUME,
+	/** @brief Represents VM state as Invalid */
 	VM_STATE_INVALID,
+	/** @brief Represents VM state as Max possible */
 	VM_STATE_MAX
 } vm_state;
 
+/** @brief Macro to create Command using function id and vmid */
 #define CREATE_CMD(func_id, vmid)	((func_id << 24U) | vmid)
 
+/** @brief
+ * This is used to get reboot/shutdown masks per VM from hypervisor.
+ * Hypervisor updates state fields on a PSCI event from the VM.
+ */
 #define QUERY_CMD			CREATE_CMD(QUERY_SYSTEM_STATE, 0)
+/** @brief Macro Represents Guest Shutdown Init State to Guest VM having vmid as VM Id */
 #define GUEST_SHUTDOWN_INIT_CMD(vmid)	CREATE_CMD(GUEST_SHUTDOWN_INIT, vmid)
+/** @brief Macro Represents Guest Shutdown Complete Command to Guest VM having vmid as VM Id */
 #define GUEST_SHUTDOWN_COMPLETE_CMD(vmid) \
 				CREATE_CMD(GUEST_SHUTDOWN_COMPLETE, vmid)
+/** @brief Macro Represents Guest Reboot Init Command to Guest VM having vmid as VM Id */
 #define GUEST_REBOOT_INIT_CMD(vmid)	CREATE_CMD(GUEST_REBOOT_INIT, vmid)
+/** @brief Macro Represents Guest Reboot Continue Command to Guest VM having vmid as VM Id */
 #define GUEST_REBOOT_CONTINUE_CMD(vmid)	CREATE_CMD(GUEST_REBOOT_CONTINUE, vmid)
+/** @brief Macro Represents Guest Reboot Complete Command to Guest VM having vmid as VM Id */
 #define GUEST_REBOOT_COMPLETE_CMD(vmid)	CREATE_CMD(GUEST_REBOOT_COMPLETE, vmid)
+/** @brief Macro Represents System Shutdown Init Command */
 #define SYS_SHUTDOWN_INIT_CMD		CREATE_CMD(SYSTEM_SHUTDOWN_INIT, 0)
+/** @brief Macro Represents System Shutdown Complete Command */
 #define SYS_SHUTDOWN_COMPLETE_CMD	CREATE_CMD(SYSTEM_SHUTDOWN_COMPLETE, 0)
+/** @brief Macro Represents System Reboot Init Command */
 #define SYS_REBOOT_INIT_CMD		CREATE_CMD(SYSTEM_REBOOT_INIT, 0)
+/** @brief Macro Represents System Reboot Complete Command */
 #define SYS_REBOOT_COMPLETE_CMD		CREATE_CMD(SYSTEM_REBOOT_COMPLETE, 0)
+/** @brief Macro Represents Guest Resume Init Command to Guest VM having vmid as VM Id */
 #define GUEST_SUSPEND_REQ_CMD(vmid)       CREATE_CMD(GUEST_SUSPEND_REQ, vmid)
+/** @brief Macro Represents Guest Suspend Init Command to Guest VM having vmid as VM Id */
 #define GUEST_SUSPEND_INIT_CMD(vmid)      CREATE_CMD(GUEST_SUSPEND_INIT, vmid)
+/** @brief Macro Represents Guest Suspend Complete Command to Guest VM having vmid as VM Id */
 #define GUEST_SUSPEND_COMPLETE_CMD(vmid) \
-		CREATE_CMD(GUEST_SUSPEND_COMPLETE, vmid)
+		CREATE_CMD(GUEST_SUSPEND_COMPLETE,vmid)
+/** @brief Macro Represents Guest Resume Init Command to Guest VM having vmid as VM Id */
 #define GUEST_RESUME_INIT_CMD(vmid)       CREATE_CMD(GUEST_RESUME_INIT, vmid)
+/** @brief Macro Represents Guest Resume Complete Command to Guest VM having vmid as VM Id */
 #define GUEST_RESUME_COMPLETE_CMD(vmid)   CREATE_CMD(GUEST_RESUME_COMPLETE, vmid)
+/** @brief Macro Represents Guest Pause Command to Guest VM having vmid as VM Id */
 #define GUEST_PAUSE_CMD(vmid)		CREATE_CMD(GUEST_PAUSE, vmid)
+/** @brief Macro Represents System Suspend Init Command */
 #define SYS_SUSPEND_INIT_CMD		CREATE_CMD(SYSTEM_SUSPEND_INIT, 0)
+/** @brief Macro Represents System Suspend Complete Command */
 #define SYS_SUSPEND_COMPLETE_CMD	CREATE_CMD(SYSTEM_SUSPEND_COMPLETE, 0)
 #define GUEST_ENTER_VM_OP_CMD		CREATE_CMD(GUEST_ENTER_VM_OP, 0)
 
+/** @} */
+
 #endif /* _TEGRA_HV_SYSMGR_H */
-- 
2.43.0

